Remove dead unwinds before drop elaboration

tmiasko opened the pull request:
As a part of drop elaboration, we identify dead unwinds, i.e., unwind
edges on a drop terminators which are known to be unreachable, because
there is no need to drop anything.

Previously, the data flow framework was informed about the dead unwinds,
and it assumed those edges are absent from MIR. Unfortunately, the data
flow framework wasn't consistent in maintaining this assumption.

In particular, if a block was reachable only through a dead unwind edge,
its state was propagated to other blocks still. This became an issue in
the context of change removes DropAndReplace terminator, since it
introduces initialization into cleanup blocks.

To avoid this issue, remove unreachable unwind edges before the drop
elaboration, and elaborate only blocks that remain reachable.

cc @Zeegomo

tmiasko commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit e80481c77ab081849f4e93f0f7e2547d9fca728a with merge 28a389c3fbb6dffe42f0f3257472197153ef82a7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e80481c77ab081849f4e93f0f7e2547d9fca728a","merge_sha":"28a389c3fbb6dffe42f0f3257472197153ef82a7"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3834031532/jobs/6527518788)
Build commit: 28a389c3fbb6dffe42f0f3257472197153ef82a7 (`28a389c3fbb6dffe42f0f3257472197153ef82a7`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3834031532/jobs/6527518788"},"merge_sha":"28a389c3fbb6dffe42f0f3257472197153ef82a7"} -->

rust-timer commented:
Queued 28a389c3fbb6dffe42f0f3257472197153ef82a7 with parent c7572670a1302f5c7e245d069200e22da9df0316, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c7572670a1302f5c7e245d069200e22da9df0316&end=28a389c3fbb6dffe42f0f3257472197153ef82a7).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (28a389c3fbb6dffe42f0f3257472197153ef82a7): [comparison URL](https://perf.rust-lang.org/compare.html?start=c7572670a1302f5c7e245d069200e22da9df0316&end=28a389c3fbb6dffe42f0f3257472197153ef82a7&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=c7572670a1302f5c7e245d069200e22da9df0316&end=28a389c3fbb6dffe42f0f3257472197153ef82a7&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c7572670a1302f5c7e245d069200e22da9df0316&end=28a389c3fbb6dffe42f0f3257472197153ef82a7&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=c7572670a1302f5c7e245d069200e22da9df0316&end=28a389c3fbb6dffe42f0f3257472197153ef82a7&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-3.6%, -3.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

Zeegomo commented on compiler/rustc_mir_transform/src/elaborate_drops.rs:
```diff
@@ -121,13 +120,27 @@ fn find_dead_unwinds<'tcx>(
+        return;
+    }
+
+    for (bb, bb_data) in body.basic_blocks_mut().iter_enumerated_mut() {
+        match bb_data.terminator_mut().kind {
+            TerminatorKind::Drop { ref mut unwind, .. }
+            | TerminatorKind::DropAndReplace { ref mut unwind, .. }
+                if unwind.is_some() && dead_unwinds.contains(bb) =>
+            {
+                *unwind = None;
+            }
+            _ => continue,
+        };
+    }
```
can't we patch the unwind target inside the loop directly?

tmiasko replied:
Data flow results cursor borrows the MIR body, so modification needs to be a separate step.

Zeegomo replied:
Ah yes, thanks.

nit: not sure how much this would impact perfs, but wouldn't iterating over `dead_unwinds` consider only the interesting parts? If we have few of those blocks we might save passing over the whole body.

Something like
```
for bb in dead_unwinds.iter() {
    match body.basic_blocks_mut()[bb].terminator_mut().kind {
      ...
   }
}
```

Zeegomo commented on compiler/rustc_mir_transform/src/elaborate_drops.rs:
```diff
@@ -38,25 +38,24 @@ impl<'tcx> MirPass<'tcx> for ElaborateDrops {
                 .iterate_to_fixpoint()
                 .into_results_cursor(body);
 
+            let reachable = traversal::reachable_as_bitset(body);
```
would it be possible to remove unreachable blocks here (or have something like `remove_dead_unwinds` as a separate pass that also prune blocks) so that we don't have to check for dead blocks when processing later?

tmiasko replied:
Yeah, it probably would be more elegant to remove dead blocks, instead of adding those extra reachability checks. The motivation behind that approach was to reuse move data between part that identifies dead unwinds and drop elaboration proper. But, maybe this isn't performance sensitive - will have to verify that.

tmiasko replied:
Due to [the cost of an additional execution MoveData::gather_moves](https://github.com/rust-lang/rust/pull/106430#issuecomment-1371749609), I reverted back to the original approach.

zeegomo reviewed (commented):
That's cool indeed, thanks!

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit dc136d1f6b41424db991463ee3ec4f0a990ee03e with merge b5371b715c5c66f73e977d7b4329bbd5bfeacb61...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dc136d1f6b41424db991463ee3ec4f0a990ee03e","merge_sha":"b5371b715c5c66f73e977d7b4329bbd5bfeacb61"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3842245002/jobs/6545058160)
Build commit: b5371b715c5c66f73e977d7b4329bbd5bfeacb61 (`b5371b715c5c66f73e977d7b4329bbd5bfeacb61`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3842245002/jobs/6545058160"},"merge_sha":"b5371b715c5c66f73e977d7b4329bbd5bfeacb61"} -->

rust-timer commented:
Queued b5371b715c5c66f73e977d7b4329bbd5bfeacb61 with parent 659e169d37990b9c730a59a96081f2ef7afbe8f1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=659e169d37990b9c730a59a96081f2ef7afbe8f1&end=b5371b715c5c66f73e977d7b4329bbd5bfeacb61).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (b5371b715c5c66f73e977d7b4329bbd5bfeacb61): [comparison URL](https://perf.rust-lang.org/compare.html?start=659e169d37990b9c730a59a96081f2ef7afbe8f1&end=b5371b715c5c66f73e977d7b4329bbd5bfeacb61&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=659e169d37990b9c730a59a96081f2ef7afbe8f1&end=b5371b715c5c66f73e977d7b4329bbd5bfeacb61&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.3%, 1.1%] | 28    |
| Regressions ❌ <br /> (secondary)  | 0.8% | [0.3%, 1.3%] | 26    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.3%, 1.1%] | 28    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=659e169d37990b9c730a59a96081f2ef7afbe8f1&end=b5371b715c5c66f73e977d7b4329bbd5bfeacb61&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=659e169d37990b9c730a59a96081f2ef7afbe8f1&end=b5371b715c5c66f73e977d7b4329bbd5bfeacb61&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

tmiasko committed:
Remove dead unwinds before drop elaboration

As a part of drop elaboration, we identify dead unwinds, i.e., unwind
edges on a drop terminators which are known to be unreachable, because
there is no need to drop anything.

Previously, the data flow framework was informed about the dead unwinds,
and it assumed those edges are absent from MIR. Unfortunately, the data
flow framework wasn't consistent in maintaining this assumption.

In particular, if a block was reachable only through a dead unwind edge,
its state was propagated to other blocks still. This became an issue in
the context of change removes DropAndReplace terminator, since it
introduces initialization into cleanup blocks.

To avoid this issue, remove unreachable unwind edges before the drop
elaboration, and elaborate only blocks that remain reachable.

Nilstrieb commented:
r? mir-opt

pnkfelix commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 4e2092bda8a1b4680ce556eb92f37240244d9e9e with merge 53cf3663aef6d77a0455af4c6e5dd1445afba7f5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4e2092bda8a1b4680ce556eb92f37240244d9e9e","merge_sha":"53cf3663aef6d77a0455af4c6e5dd1445afba7f5"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4016555176/jobs/6902875493)
Build commit: 53cf3663aef6d77a0455af4c6e5dd1445afba7f5 (`53cf3663aef6d77a0455af4c6e5dd1445afba7f5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4016555176/jobs/6902875493"},"merge_sha":"53cf3663aef6d77a0455af4c6e5dd1445afba7f5"} -->

rust-timer commented:
Queued 53cf3663aef6d77a0455af4c6e5dd1445afba7f5 with parent 40fda7b3fe2b10c6e1a0568b59516f5e7f381886, future [comparison URL](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=53cf3663aef6d77a0455af4c6e5dd1445afba7f5).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (53cf3663aef6d77a0455af4c6e5dd1445afba7f5): [comparison URL](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=53cf3663aef6d77a0455af4c6e5dd1445afba7f5&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=53cf3663aef6d77a0455af4c6e5dd1445afba7f5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-1.4%, -0.2%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=53cf3663aef6d77a0455af4c6e5dd1445afba7f5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 6.0%  | [5.0%, 7.6%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.0%, -1.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=53cf3663aef6d77a0455af4c6e5dd1445afba7f5&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

tmiasko committed:
Emit diff instead of after mir in ElaborateDrops tests

to make it easy to understand chnages made by elaboration.

tmiasko committed:
./x.py test --bless

cjgillot commented:
Could you make the ElaborateDrops tests output ElaborateDrops.diff instead of ElaborateDrops.after.mir? It will be easier to check the output.

Sorry for the slow review. r=me after a rebase.

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 2a70524397728e4d13e494667fc4865af42c879b has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 2a70524397728e4d13e494667fc4865af42c879b -->
<!-- homu: {"type":"Approved","sha":"2a70524397728e4d13e494667fc4865af42c879b","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented:
Will this conflict with https://github.com/rust-lang/rust/pull/107732? Borrowck needs to see those dead unwinds, so if drop elaboration happens before borrowck, removing dead unwinds will change behavior, right?

tmiasko commented:
The dead unwinds are removed by the elaboration regardless. If they are indeed necessary then the conflict is inherent to execution of drop elaboration before the borrowck. Probably something to note in #107732 (I am not familiar with borrowck, to say if this is an issue or not).

bors commented:
:hourglass: Testing commit 2a70524397728e4d13e494667fc4865af42c879b with merge d87a5398c8f6182df9c4a6e6628ccfcb8aec0481...
<!-- homu: {"type":"BuildStarted","head_sha":"2a70524397728e4d13e494667fc4865af42c879b","merge_sha":"d87a5398c8f6182df9c4a6e6628ccfcb8aec0481"} -->

ehuss commented:
@bors retry

aarch64-gnu was offline, and according to Zulip should be back up.

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4256925860/jobs/7408842341) [(plain)](https://github.com/rust-lang-ci/rust/commit/d87a5398c8f6182df9c4a6e6628ccfcb8aec0481/checks/11564273685/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit 2a70524397728e4d13e494667fc4865af42c879b with merge 7d68ab0858e4dc6a33bc9a237287b61eb2f99671...
<!-- homu: {"type":"BuildStarted","head_sha":"2a70524397728e4d13e494667fc4865af42c879b","merge_sha":"7d68ab0858e4dc6a33bc9a237287b61eb2f99671"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

WaffleLapkin commented:
@bors retry

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4259715314/jobs/7419920238) [(plain)](https://github.com/rust-lang-ci/rust/commit/7d68ab0858e4dc6a33bc9a237287b61eb2f99671/checks/11578463113/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit 2a70524397728e4d13e494667fc4865af42c879b with merge e08ad60cc85327cd0f7eadf1450b622cd2a0774e...
<!-- homu: {"type":"BuildStarted","head_sha":"2a70524397728e4d13e494667fc4865af42c879b","merge_sha":"e08ad60cc85327cd0f7eadf1450b622cd2a0774e"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4267978453/jobs/7430186749)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4267978453/jobs/7430186749","builder_name":"checks-actions"} -->

ehuss commented:
@bors retry

#108227

rust-log-analyzer commented:
The job **`x86_64-mingw-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4267978453/jobs/7430114850) [(plain)](https://github.com/rust-lang-ci/rust/commit/e08ad60cc85327cd0f7eadf1450b622cd2a0774e/checks/11591913458/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[2541/3025] Linking CXX static library lib\libLLVMSymbolize.a
[2542/3025] Linking CXX static library lib\libLLVMObjectYAML.a
[2543/3025] Linking CXX static library lib\libLLVMDebuginfod.a
[2544/3025] Linking CXX static library lib\libLLVMProfileData.a
FAILED: lib/libLLVMProfileData.a 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E rm -f lib\libLLVMProfileData.a && D:\a\rust\rust\mingw64\bin\ar.exe qc lib\libLLVMProfileData.a  lib/ProfileData/CMakeFiles/LLVMProfileData.dir/GCOV.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/InstrProf.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/InstrProfCorrelator.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/InstrProfReader.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/InstrProfWriter.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/MemProf.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/ProfileSummaryBuilder.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/SampleProf.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/SampleProfReader.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/SampleProfWriter.cpp.obj lib/ProfileData/CMakeFiles/LLVMProfileData.dir/RawMemProfReader.cpp.obj && D:\a\rust\rust\mingw64\bin\ranlib.exe lib\libLLVMProfileData.a && cd ."
D:\a\rust\rust\mingw64\bin\ar.exe: could not create temporary file whilst writing archive: no more archived files
[2545/3025] Building CXX object lib/XRay/CMakeFiles/LLVMXRay.dir/Profile.cpp.obj
[2546/3025] Building CXX object lib/XRay/CMakeFiles/LLVMXRay.dir/RecordInitializer.cpp.obj
[2547/3025] Building CXX object lib/XRay/CMakeFiles/LLVMXRay.dir/LogBuilderConsumer.cpp.obj
[2548/3025] Building CXX object lib/WindowsManifest/CMakeFiles/LLVMWindowsManifest.dir/WindowsManifestMerger.cpp.obj
[2548/3025] Building CXX object lib/WindowsManifest/CMakeFiles/LLVMWindowsManifest.dir/WindowsManifestMerger.cpp.obj
[2549/3025] Building CXX object utils/FileCheck/CMakeFiles/FileCheck.dir/FileCheck.cpp.obj
[2550/3025] Building CXX object utils/PerfectShuffle/CMakeFiles/llvm-PerfectShuffle.dir/PerfectShuffle.cpp.obj
[2551/3025] Building CXX object lib/WindowsDriver/CMakeFiles/LLVMWindowsDriver.dir/MSVCPaths.cpp.obj
ninja: build stopped: subcommand failed.
command did not execute successfully, got: exit code: 1


build script failed, must exit now', C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cmake-0.1.48\src\lib.rs:975:5
 finished in 220.609 seconds
Build completed unsuccessfully in 0:06:00
Build completed unsuccessfully in 0:06:00
make: *** [Makefile:78: ci-mingw-subset-1] Error 1
```

</details>

bors commented:
:hourglass: Testing commit 2a70524397728e4d13e494667fc4865af42c879b with merge 313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb...
<!-- homu: {"type":"BuildStarted","head_sha":"2a70524397728e4d13e494667fc4865af42c879b","merge_sha":"313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4268692929/jobs/7432420073)
Approved by: cjgillot
Pushing 313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4268692929/jobs/7432420073"},"merge_sha":"313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb"} -->

rust-timer commented:
Finished benchmarking commit (313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb): [comparison URL](https://perf.rust-lang.org/compare.html?start=6ffabf3c8f2ed81d0248c54365445f28bd9507e5&end=313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6ffabf3c8f2ed81d0248c54365445f28bd9507e5&end=313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6ffabf3c8f2ed81d0248c54365445f28bd9507e5&end=313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0% | [3.0%, 3.0%] | 1     |
| Regressions ❌ <br /> (secondary)  | 4.7% | [4.7%, 4.7%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.0% | [3.0%, 3.0%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6ffabf3c8f2ed81d0248c54365445f28bd9507e5&end=313219c0b6bacabe0fc72a8eb3912b8fc08b9bdb&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 5.1% | [3.9%, 6.2%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>
